### PR TITLE
[4.0] Api components

### PR DESCRIPTION
**Pull request by o-samoilov**
~~Pull Request for Issue #26002.~~

### Summary of Changes

Implemented Joomla webservices:

* com_banners
* com_categories
* com_config
* com_contact
* com_content
* com_contenthistory
* com_fields
* com_installer
* com_languages
* com_menus
* com_messages
* com_modules
* com_newsfeeds
* com_plugins
* com_privacy
* com_redirect
* com_tags
* com_templates
* com_users

Documentation:
https://docs.joomla.org/J4.x:Joomla_Core_APIs

### CODE REVIEW

File: tests/Codeception/api/com_content/ContentCest.php [hunk 1]
```diff
@@ -71,24 +71,24 @@ public function testCrudOnArticle(ApiTester $I)
 			'alias' => 'tobias'
 		];
 
-		$I->sendPOST('/article', $testarticle);
+		$I->sendPOST('/content/article', $testarticle);
```
Comment (Hackwar) [inline tests/Codeception/api/com_content/ContentCest.php:74]: Shouldn't this (and all subsequent test routes) be `/v1/content/article...`?

File: api/components/com_banners/View/Banners/JsonapiView.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+/**
+ * The banners view
+ *
+ * @since  4.0.0
+ */
+class JsonapiView extends BaseApiView
+{
+	/**
+	 * The fields to render item in the documents
```
Comment (brianteeman) [inline api/components/com_banners/View/Banners/JsonapiView.php]: This comment description or the one below can not be correct as they are the same but describe different functions.

File: api/components/com_banners/View/Banners/JsonapiView.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+defined('_JEXEC') or die;
+
+use Joomla\CMS\MVC\View\JsonApiView as BaseApiView;
+
+/**
+ * The banners view
+ *
+ * @since  4.0.0
+ */
+class JsonapiView extends BaseApiView
```
Comment (brianteeman) [inline api/components/com_banners/View/Banners/JsonapiView.php]: BaseApiView has capital BAV
Can the name for JsonapiView not be named in the same style as well?
eg JsonApiView

Comment (brianteeman) [reply api/components/com_banners/View/Banners/JsonapiView.php]: especially as it is `use Joomla\CMS\MVC\View\JsonApiView as BaseApiView;
`

File: api/components/com_categories/Controller/CategoriesController.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
+<?php
+/**
+ * @package     Joomla.API
+ * @subpackage  com_newsfeeds
```
Comment (brianteeman) [inline api/components/com_categories/Controller/CategoriesController.php]:
```suggestion
 * @subpackage  com_categories
```

File: api/components/com_categories/View/Categories/JsonapiView.php [hunk 5]
```diff
@@ -0,0 +1,163 @@
+<?php
+/**
+ * @package     Joomla.API
+ * @subpackage  com_newsfeeds
```
Comment (brianteeman) [inline api/components/com_categories/View/Categories/JsonapiView.php]:
```suggestion
 * @subpackage  com_categories
```

File: api/components/com_config/Controller/ApplicationController.php [hunk 6]
```diff
@@ -0,0 +1,154 @@
+<?php
+/**
+ * @package     Joomla.API
+ * @subpackage  com_plugins
```
Comment (brianteeman) [inline api/components/com_config/Controller/ApplicationController.php]:
```suggestion
 * @subpackage  com_config
```

File: api/components/com_config/Controller/ComponentController.php [hunk 7]
```diff
@@ -0,0 +1,165 @@
+<?php
+/**
+ * @package     Joomla.API
+ * @subpackage  com_plugins
```
Comment (brianteeman) [inline api/components/com_config/Controller/ComponentController.php]:
```suggestion
 * @subpackage  com_config
```

File: api/components/com_installer/View/Overrides/JsonapiView.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
+<?php
+/**
+ * @package     Joomla.API
+ * @subpackage  com_content
```
Comment (brianteeman) [inline api/components/com_installer/View/Overrides/JsonapiView.php]:
```suggestion
 * @subpackage  com_installer
```

File: api/components/com_languages/View/Languages/JsonapiView.php [hunk 9]
```diff
@@ -0,0 +1,89 @@
+<?php
+/**
+ * @package     Joomla.API
+ * @subpackage  com_content
```
Comment (brianteeman) [inline api/components/com_languages/View/Languages/JsonapiView.php]:
```suggestion
 * @subpackage  com_languages
```

File: api/components/com_languages/View/Overrides/JsonapiView.php [hunk 10]
```diff
@@ -0,0 +1,103 @@
+<?php
+/**
+ * @package     Joomla.API
+ * @subpackage  com_content
```
Comment (brianteeman) [inline api/components/com_languages/View/Overrides/JsonapiView.php]:
```suggestion
 * @subpackage  com_languages
```

File: api/components/com_languages/View/Strings/JsonapiView.php [hunk 11]
```diff
@@ -0,0 +1,101 @@
+<?php
+/**
+ * @package     Joomla.API
+ * @subpackage  com_content
```
Comment (brianteeman) [inline api/components/com_languages/View/Strings/JsonapiView.php]:
```suggestion
 * @subpackage  com_languages
```

File: api/components/com_users/View/Users/JsonapiView.php [hunk 12]
```diff
@@ -0,0 +1,115 @@
+<?php
+/**
+ * @package     Joomla.API
+ * @subpackage  com_redirect
```
Comment (brianteeman) [inline api/components/com_users/View/Users/JsonapiView.php]:
```suggestion
 * @subpackage  com_users
```

File: tests/Codeception/api/com_content/ContentCest.php (hunk 1 above)
Comment (Hackwar) [reply tests/Codeception/api/com_content/ContentCest.php:74]: No, it shouldn't. That is handled in the test configuration. Which I have to say is pretty confusing to me...

File: api/components/com_banners/View/Banners/JsonapiView.php (hunk 3 above)
Comment (o-samoilov) [reply api/components/com_banners/View/Banners/JsonapiView.php]: Yes, you're right, thanks :)

File: tests/Codeception/api/com_content/ContentCest.php (hunk 1 above)
Comment (o-samoilov) [reply tests/Codeception/api/com_content/ContentCest.php:74]: Yes, this is in the configs, maybe when writing tests, will need to revise it

File: installation/sql/mysql/joomla.sql [hunk 13]
```diff
@@ -697,7 +697,22 @@ INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`,
 (0, 'plg_extension_finder', 'plugin', 'finder', 'extension', 0, 1, 1, 0, '', '', 0, '0000-00-00 00:00:00', 0, 0),
 (0, 'plg_system_skipto', 'plugin', 'skipto', 'system', 0, 1, 1, 0, '', '{}',  0, '0000-00-00 00:00:00', 0, 0),
 (0, 'plg_api-authentication_basic', 'plugin', 'basic', 'api-authentication', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
+(0, 'plg_webservices_banners', 'plugin', 'banners', 'webservices', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
+(0, 'plg_webservices_contact', 'plugin', 'contact', 'webservices', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0,  0),
+(0, 'plg_webservices_config', 'plugin', 'config', 'webservices', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0,  0),
 (0, 'plg_webservices_content', 'plugin', 'content', 'webservices', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
+(0, 'plg_webservices_fields', 'plugin', 'fields', 'webservices', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0,  0),
```
Comment (wilsonge) [inline installation/sql/mysql/joomla.sql]: This needs to be removed from here and the ini files (it got bundled with the components)

File: administrator/components/com_templates/Model/StylesModel.php [hunk 14]
```diff
@@ -61,15 +62,19 @@ public function __construct($config = array(), MVCFactoryInterface $factory = nu
...
-		$this->setState('filter.search', $this->getUserStateFromRequest($this->context . '.filter.search', 'filter_search', '', 'string'));
-		$this->setState('filter.template', $this->getUserStateFromRequest($this->context . '.filter.template', 'filter_template', '', 'string'));
-		$this->setState('filter.menuitem', $this->getUserStateFromRequest($this->context . '.filter.menuitem', 'filter_menuitem', '', 'cmd'));
-
-		// Special case for the client id.
-		$clientId = (int) $this->getUserStateFromRequest($this->context . '.client_id', 'client_id', 0, 'int');
-		$clientId = (!in_array($clientId, array (0, 1))) ? 0 : $clientId;
-		$this->setState('client_id', $clientId);
+		$app = Factory::getApplication();
+		if (!$app->isClient('api'))
```
Comment (Quy) [inline administrator/components/com_templates/Model/StylesModel.php:67]:
```suggestion

		if (!$app->isClient('api'))
```

File: administrator/components/com_templates/Model/StylesModel.php [hunk 15]
```diff
@@ -61,15 +62,19 @@ public function __construct($config = array(), MVCFactoryInterface $factory = nu
...
+		{
+			// Load the filter state.
+			$this->setState('filter.search', $this->getUserStateFromRequest($this->context . '.filter.search', 'filter_search', '', 'string'));
+			$this->setState('filter.template', $this->getUserStateFromRequest($this->context . '.filter.template', 'filter_template', '', 'string'));
+			$this->setState('filter.menuitem', $this->getUserStateFromRequest($this->context . '.filter.menuitem', 'filter_menuitem', '', 'cmd'));
+
+			// Special case for the client id.
+			$clientId = (int) $this->getUserStateFromRequest($this->context . '.client_id', 'client_id', 0, 'int');
+			$clientId = (!in_array($clientId, array (0, 1))) ? 0 : $clientId;
```
Comment (Quy) [inline administrator/components/com_templates/Model/StylesModel.php]:
```suggestion
			$clientId = !in_array($clientId, [0, 1]) ? 0 : $clientId;
```

File: administrator/language/en-GB/en-GB.plg_webservices_banners.sys.ini [hunk 16]
```diff
@@ -0,0 +1,7 @@
+; Joomla! Project
+; Copyright (C) 2005 - 2019 Open Source Matters. All rights reserved.
+; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
+; Note : All ini files need to be saved as UTF-8
+
+PLG_WEBSERVICES_BANNERS="Web Services - Banners"
+PLG_WEBSERVICES_BANNERS_XML_DESCRIPTION="Used to add banners routes to the API for your website."
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.plg_webservices_banners.sys.ini]:
```suggestion
PLG_WEBSERVICES_BANNERS_XML_DESCRIPTION="Add banners routes to the API for your website."
```

Comment (Quy) [reply administrator/language/en-GB/en-GB.plg_webservices_banners.sys.ini]: Repeat in the other files.

File: api/components/com_banners/View/Banners/JsonApiView.php [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+		'ordering',
+		'purchase_type',
+		'language',
+		'publish_up',
+		'publish_down',
+		'language_image',
+		'editor',
+		'category_title',
+		'client_name',
+		'client_purchase_type'
```
Comment (Quy) [inline api/components/com_banners/View/Banners/JsonApiView.php]:
```suggestion
		'client_purchase_type',
```

File: api/components/com_banners/View/Clients/JsonApiView.php [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+		'checked_out_time',
+		'state',
+		'metakey',
+		'purchase_type',
+		'nbanners',
+		'editor',
+		'count_published',
+		'count_unpublished',
+		'count_trashed',
+		'count_archived'
```
Comment (Quy) [inline api/components/com_banners/View/Clients/JsonApiView.php]:
```suggestion
		'count_archived',
```

File: api/components/com_categories/View/Categories/JsonApiView.php [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+		'language',
+		'language_title',
+		'language_image',
+		'editor',
+		'access_level',
+		'author_name',
+		'count_trashed',
+		'count_unpublished',
+		'count_published',
+		'count_archived'
```
Comment (Quy) [inline api/components/com_categories/View/Categories/JsonApiView.php]:
```suggestion
		'count_archived',
```

File: api/components/com_categories/View/Categories/JsonApiView.php [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
```
Comment (Quy) [inline api/components/com_categories/View/Categories/JsonApiView.php]:
```suggestion
		'count_archived',
```

File: api/components/com_config/Controller/ComponentController.php [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+				{
+					$messages[] = "{$errors[$i]}";
+				}
+			}
+
+			throw new InvalidParameterException(implode("\n", $messages));
+		}
+
+		// Attempt to save the configuration.
+		$data = array(
```
Comment (Quy) [inline api/components/com_config/Controller/ComponentController.php]:
```suggestion
		$data = [
```

File: api/components/com_config/Controller/ComponentController.php [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+			'params' => $validData,
+			'id'     => ExtensionHelper::getExtensionRecord($option)->extension_id,
+			'option' => $option
+		);
```
Comment (Quy) [inline api/components/com_config/Controller/ComponentController.php]:
```suggestion
		];
```

File: api/components/com_config/View/Application/JsonApiView.php [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+
+namespace Joomla\Component\Config\Api\View\Application;
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\Extension\ExtensionHelper;
+use Joomla\CMS\MVC\View\JsonApiView as BaseApiView;
+use Joomla\CMS\Serializer\JoomlaSerializer;
+use Joomla\CMS\Uri\Uri;
+use Tobscure\JsonApi\Collection;
```
Comment (Quy) [inline api/components/com_config/View/Application/JsonApiView.php:18]: Sort in alpha order.

File: administrator/components/com_menus/Model/ItemModel.php [hunk 24]
```diff
@@ -957,15 +957,23 @@ protected function populateState()
...
 
-		if (!($parentId = $app->getUserState('com_menus.edit.item.parent_id')))
+		if (!$app->isClient('api'))
+		{
+			$parentId = $app->getUserState('com_menus.edit.item.parent_id');
+			$menuType = $app->getUserStateFromRequest('com_menus.items.menutype', 'menutype', '', 'string');
+		}
+		else
+		{
+			$menuType = $app->input->get('com_menus.items.menutype');
```
Comment (wilsonge) [inline administrator/components/com_menus/Model/ItemModel.php:960]: We need to set `parentId` to null here I guess else it's undefined for the next line

**Comment by wilsonge** [comment]
Thankyou very much for all your hard work @a-samoylov . Very much appreciated

**Pull request merged**